pkg/controller/sparkapplication/controller.go

```diff
@@ -289,397 +298,471 @@ func (c *Controller) processNextItem() bool {
-			"SparkApplicationSubmissionFailed",
-			"SparkApplication %s failed submission: %s",
-			update.name,
-			update.errorMessage)
-
-		if shouldRetrySubmission(app) {
-			submissionRetries++
-			c.handleResubmission(app, submissionRetries)
-		} else {
-			glog.Infof("Not retrying submission of SparkApplication %s", update.name)
-		}
+	// Retry if we have waited at-least equal to attempts*RetryInterval since we do a linear back-off.
+	interval := time.Duration(*retryInterval) * time.Second * time.Duration(attemptsDone)
+	currentTime := time.Now()
+	if currentTime.After(lastEventTime.Add(interval)) {
+		return true
 	}
-
-	return c.updateSparkApplicationStatusWithRetries(app, func(status *v1alpha1.SparkApplicationStatus) {
-		status.AppState.State = update.state
-		status.AppState.ErrorMessage = update.errorMessage
-		status.SubmissionRetries = submissionRetries
-		if !update.submissionTime.IsZero() {
-			status.SubmissionTime = update.submissionTime
-		}
-	})
+	return false
 }
 
-func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdate) *v1alpha1.SparkApplication {
-	glog.V(2).Infof(
-		"Received state update of executor %s for SparkApplication %s in namespace %s with state %s",
-		update.executorID, update.appName, update.appNamespace, update.state)
-
-	app, err := c.getSparkApplication(update.appNamespace, update.appName)
+// submitSparkApplication creates a new submission for the given SparkApplication and submits it using spark-submit.
+func (c *Controller) submitSparkApplication(app *v1alpha1.SparkApplication) *v1alpha1.SparkApplication {
+	submissionCmdArgs, err := buildSubmissionCommandArgs(app)
 	if err != nil {
-		// Update may be the result of pod deletion due to deletion of the owning SparkApplication object.
-		// Ignore the error if the owning SparkApplication object does not exist.
-		if !errors.IsNotFound(err) {
-			glog.Errorf("failed to get SparkApplication %s in namespace %s from the store: %v", update.appName,
-				update.appNamespace, err)
+		app.Status = v1alpha1.SparkApplicationStatus{
+			AppState: v1alpha1.ApplicationState{
+				State:        v1alpha1.FailedSubmissionState,
+				ErrorMessage: err.Error(),
+			},
+			SubmissionAttempts:        app.Status.SubmissionAttempts + 1,
+			LastSubmissionAttemptTime: metav1.Now(),
 		}
-		return nil
+		return app
 	}
 
-	// The controller may receive status updates of executors belonging to the previous run of the application and
-	// should ignore such updates. For example, if a user makes some change to a SparkApplication and does 'kube apply'
-	// to update the SparkApplication object on the API server while the application is running, the controller may
-	// still receive status updates of executors of the old application run even after the new run starts.
-	if update.appID != app.Status.AppID {
-		return nil
+	// Try submitting App.
+	submission := newSubmission(submissionCmdArgs, app)
+	sparkHome, present := os.LookupEnv(sparkHomeEnvVar)
```

Comment:
Can we put the code for executing the `spark-submit` command in a utility function like what we currently have?

Reply:
Done, created a helper method in submission.go